### ISSUE TITLE
Skip some Amp tests

### ISSUE THREAD

**Issue by davidwdan** [WyriHaximus/reactphp-async-interop-loop#16]
I think that these two tests are out of our control:

https://travis-ci.org/WyriHaximus/reactphp-async-interop-loop/jobs/189426071#L660
https://travis-ci.org/WyriHaximus/reactphp-async-interop-loop/jobs/189426071#L673

Should they be skipped?

**Comment by kelunik** [comment]
These should still pass. @trowski @bwoebi

**Comment by WyriHaximus** [comment]
Agreed, those should still pass and we shouldn't dismiss then just yet

**Comment by mbonneau** [comment]
Amp calls the read callback 3 times instead of 2 - this seems like it is testing an internal thing with react that Amp doesn't necessarily follow. Is it a failure that Amp notifies the callback 3 times instead of 2 times? There is still data in the read buffer all 3 times.

Or is it a requirement that when `tick` is called it must only call the read callback once?

**Comment by kelunik** [comment]
@mbonneau I'm not sure, but I expected one tick to call the callback only once.

**Comment by mbonneau** [comment]
I tested this code which simulates the call only 2 times test from react:
```php
<?php

// using amphp/amp commit 9ff73ac6816eb63562fa63a091410f8d7404f4dc (dev-master)

use Interop\Async\Loop;

require_once __DIR__ . '/vendor/autoload.php';

function tick() {
    Loop::execute(function () {
        Loop::stop();
    }, Loop::get());
}

$stream = fopen('php://temp', 'r+');

Loop::onReadable($stream, function () {
    echo "Callable called...\n";
});

fwrite($stream, "abc");
tick();

fwrite($stream, "123");
tick();
```
I get output of:
```
Callable called...
Callable called...
Callable called...
```
Which again - I don't know if this is actually an issue because there is data in the buffer all 3 times - but yeah, I also would expect it to only be called 2 times.

**Comment by trowski** [comment]
The `onReadable` watcher is invoked 3 times because a `defer` watcher is added before starting the loop here: https://github.com/WyriHaximus/reactphp-async-interop-loop/blob/master/src/AsyncInteropLoop.php#L169-L175

Since watchers are only activated at the end of of a tick and the read watcher remains active between calling `run()` and `stop()`, Since calling `stop()` does not immediately stop the loop, it finishes executing the current tick, three ticks of the loop will be executed in total.

**Comment by mbonneau** [comment]
In the example above - I use `execute` instead of `defer` and get the same results.

Is there a way to make this only call twice so that behavior matches and we wouldn't have to skip the tests?

**Comment by bwoebi** [comment]
@trowski I think this isn't spec compliant then.

The spec says:

> The deferred callable MUST be executed in the next tick of the event loop

The next tick is when run() is called the next time.

I think I should add a test to event-loop-test then?

**Comment by bwoebi** [comment]
@mbonneau I think it's an implementation bug in amp/loop. … Let's see what @trowski says?

**Comment by trowski** [comment]
@mbonneau `run()` calls `execute()`, so actually it's the same.

@bwoebi Yes, we should add some tests to `event-loop-test`. Amp currently activates watchers at the end of each tick rather than the beginning (which is equivalent in a running loop). I'll update the Amp implementation (which is a trivial change) and you can rerun the tests.

Maybe some wording in the spec that watchers should be activated at the beginning of each tick (but not during the tick)?

**Comment by mbonneau** [comment]
@trowski @bwoebi - Thanks - I will keep an eye out for the update and check it again.

**Comment by WyriHaximus** [comment]
Just updated and kicked off new tests running at https://travis-ci.org/WyriHaximus/reactphp-async-interop-loop/builds/189555034

**Comment by bwoebi** [comment]
@trowski The Loop::enable() docblock specifies:

> Watchers (enabling or new watchers) MUST immediately be marked as enabled, but only be activated (i.e. callbacks can be called) right before the next tick

Which is exactly what you asked for - so nothing to be done there.

**Comment by kelunik** [comment]
I think that should be moved to the class level doc block instead.

**Comment by mbonneau** [comment]
The example above does only call the callback 2 times and it looks like that tests are working.

Can this be closed?

**Comment by bwoebi** [comment]
Yes, looks like it's working now.

**Comment by trowski** [comment]
@bwoebi "Activated right before the next tick" still leaves the possibility of activating at the end of the tick, as that will be before the next tick. Activating at the end of the tick was the correct behavior to me as it left watchers activated if the loop was stopped after that tick, putting it in a state ready to execute the next tick. It would be better if it said "activated at the beginning of the next tick."

**Comment by bwoebi** [comment]
I don't agree that the wording implies that, but if you think your proposed wording is less ambiguous, then go ahead and commit it.

**Comment by kelunik** [comment]
@trowski A test should cover that assumption and break it.

**Comment by trowski** [comment]
@kelunik Yep, which didn't exist before, added here: https://github.com/async-interop/event-loop-test/blob/master/src/Test.php#L1217-L1240 :smile:
